[DOCS] explain release cycle in branches (#2475)
DivanteLtd/vue-storefront: docs/guide/basics/release-cycle.md

- v1.x-RC.y - unstable version with cutting-edge features ready to test and feedback
- v1.x.y - stable version of the software **ready for production use**.

## How new features are merged

During RC features Pull Request with new features after feedback and acceptance are normally merged to `develop` branch. 
After entering the Stabilization Phase we are tagging current develop branch, creating a `RC-x` (where `x` is a number of current version) branch from it and working on stabilization there. 
During the stabilization phase new features are merged to develop branch and will be merged on next `RC` phase.

## Release cycle flow

### 1. Development phase

> The first phase of cycle we're mostly focusing on features and improvements. Branches in that phase should be created from actual `develop`, also PRs should be pointed to this branch. Changes merged to develop are available to test on https://test.storefrontcloud.io

Reviewer: I think we should wrap all branches names in the same way.
```suggestion
The first phase of cycle we're mostly focusing on features and improvements. Branches in that phase should be created from actual `develop`, also PRs should be pointed to this branch. Changes merged to `develop` are available to test on https://test.storefrontcloud.io
```

![Development phase](./assets/release-cycle-1.png)

### 2. Release Candidate phase

> At some point, when milestone for next minor versions is completed, we're creating new branch from `develop` called `RC-x.y` (example: `rc-1.9`).

Reviewer: I'd consider following Git-flow naming, so:
- `release/` prefix
- name of the branch is used to create a tag after merging (git-flow tools are doing this automatically)

So it will be `release/v1.9.0-rc.1`

Reviewer: If it's okay we should link to some git-flow description and update the "How Vue Storefront versions are released" part too.

Author: that's great, i'll change that

Author: actually - with rc?
So what with fixes?

Idea was to have one RC branch, and tag it as next RC versions. And now we'll have multiple rc branches per single release; i think this would be confusing. Could you elaborate more how is that better?

Reviewer 2: I guess having too many branches is a bad idea. One rc which is equvallent of current develop is ok. This tagging is not giving us any advantage

Reviewer: That might be true, will be easier to target changes.
So the final version will be:
`release/v1.9.0-rc` - just on for the whole stablization phase
`hotfix/v1.9.1` - new branch for each hotfix of stable release
am I correct?

> At some point, when milestone for next minor versions is completed, we're creating new branch from `develop` called `RC-x.y` (example: `rc-1.9`).

Reviewer: Letters size (case) is not consistent.
```suggestion
At some point, when milestone for next minor versions is completed, we're creating new branch from `develop` called `rc-x.y` (example: `rc-1.9`).
```

> After that new branch is tagged as first RC for version (example `v1.9.0-RC.1`). Then it's ready for testing by community.

Reviewer: I think it's required to use lowercase.
```suggestion
After that new branch is tagged as first RC for version (example `v1.9.0-rc.1`). Then it's ready for testing by community.
```

> During tests, feedback and stabilisation there could be multiple RC releases on this branch. When improvement is made on this phase, then branch should be created from actual `rc-x.y` and should not contain features at this point - only improvements for current release.

Reviewer: ```suggestion
During tests, feedback and stabilization there could be multiple Release Candidate versions on this branch. When improvement is made on this phase, then branch should be created from actual `rc-x.y` and should not contain features at this point - only improvements for current release.
```

> When release branch is updated, then should `develop` should be synchronized with it.

Reviewer: I think the `develop` should be updated during (any) version release, same as in git-flow.

So `release/*` branch should be merged simultaneously to `master` and `develop`.

Author: i disagree with that. Release branch should be merged only to develop. To master should be merged only to change RC to stable version.
So:
- changes in release branch(stabilization fixes) -> merge to develop
- changes in master branch(hotfix) -> merge to release branch and develop branch

Reviewer 2: The  idea is to have
- `master` should be always a one with stable, production ready code. That means only releases from stabilization phase and hotfixes are applied there
- `develop` is a working branch and should be always up to date with latest code no matter if we are on RC or STABILIZATION phase

Reviewer: So how to release RC version? Tag RC releases on this temporary branch? I don't think it's a good idea.

I know that probably lots of people are just based on `master` branch and don't give a f about tags, but maybe they should start :P

Reviewer: ```suggestion
After merging a set of bugfixes and improvements into `release/*` branch, it needs to be tagged as the next RC version and merged into the `develop` branch to update it.
```

Reviewer: So if we want to tag on this temporary branch, we should make it clear (also there is now "should" two times and in general this senstence is kinda weird :D )

```suggestion
After merging a set of bugfixes and improvements into `release/*` branch, it needs to be tagged as the next RC version and merged into the `develop` branch, to update it.
```

![Release Candidate phase](./assets/release-cycle-2.png)

### 3. Release phase

> When RC version is stable, then release branch is merged into `master` and tagged as next stable version (example `v1.9.0`). After that currently merged release branch is deleted and starts new development phase. Yet, it some critical bug is discovered on current stable version, then new branch should be created from actual `master` and merged into `hotfix-x.y.z` (where `z` is next path intertion; example: `hotfix-1.9.1`). After merging and testing hotfixes, this branch (like release branch before) is merged into `master` and tag by next path version from hotfix branch name.

Reviewer: Same as releases, I think we should stick with git-flow naming and just refer to semantic versioning description, instead of trying to describe here what the patch version is.
```suggestion
When RC version is stable, then release branch is merged into `master` and tagged as next stable version (example `v1.9.0`). After that currently merged release branch is deleted and starts new development phase. If some critical bug is discovered on current stable version, then new hotfix (for example `hotfix/v1.9.1`, following [Semantic Versioning rules](https://semver.org/)) branch should be created from actual `master`. After merging and testing hotfixes, this branch (like release branch before) is merged into `master` and tagged following [Semantic Versioning rules](https://semver.org/).
```

![Release phase](./assets/release-cycle-3.png)

### Summary

> Important thing to notice is that this releasing cycle ensures stable and reliable Storefront releases. Phases are also not blocking new features - you can develop new features on any phase, but it should be merged only to develop and go through whole cycle.

Reviewer: ```suggestion
Important thing to notice is that this releasing cycle ensures stable and reliable Storefront releases. Phases are also not blocking new features - you can develop new features on any phase, but it should be merged only to `develop` branch and go through whole cycle.
```
